### DISCUSSION TITLE
Some corner use cases

### DISCUSSION THREAD

**Discussion by zcbenz** [nodejs/single-executable#30] [General]
On Windows, extracting DLLs and executing them is a red alarm to anti-virus softwares, so when supporting native addons with VFS there should be a way to leave the DLLs on the real filesystem, like the ["unpack" feature of ASAR](https://github.com/electron/asar#excluding-multiple-resources-from-being-packed).

Another usage of the "unpack" feature is to feed assets to APIs outside JS world. Most system APIs and third party C libraries only read files on the real filesystem, for example when creating a tray icon you have to pass the path of the icon on the real filesystem. You can of course extract the file before passing its path, but that would require modifications to existing code to support VFS.

---

On macOS executables are usually signed and the single executable implementation should be compatible with signing. Many single executable implementations are concatenating the packages to the tail of the executables, which would be broken after signing, but it can be fixed by [including payload in str table](https://github.com/vercel/pkg/pull/1164).

---

While this repo is about single executable, it would be nice supporting reading from multiple bundled packages in one process, so it would be possible to distribute Node.js apps as single files. For example users can execute downloaded apps with something like `node downloaded_app.nodeapp` or by double clicking the package file, and app developers can do `fork('/path/to/another/app.nodeapp')`.

---

There should be a way to manipulate the bundled package file itself, which is important for creating toolings. Electron does it with [`original-fs` module](https://github.com/electron/electron/blob/v12.0.0/docs/tutorial/application-packaging.md#treating-an-asar-archive-as-a-normal-file).

---

When supporting VFS in Node's builtin modules, instead of implementing it in JS, I think it is better implementing in C and make the code reusable for other libuv users. So it would be easier for other runtimes like deno and bun to adopt the support.

**Comment by cspotcode** [comment]
> ...there should be a way to leave the DLLs on the real filesystem

Should this be included in a formal list of design requirements, for the VFS or for  the SPA bootstrapper?

- built-in support for a manifest describing which files to extract
- SPA will automatically extract files when it starts as needed
  - can use a temp cache if already exists?  not sure
- JS APIs allow the SPA to get the paths to these extracted assets
  - These paths will be different from the VFS paths; will not share a common base path
  - Can/should the VFS have virtual "symlinks" to the extracted paths on the real FS?  For example, a symlink from `/foo/spa-exe/__vfs__/extracted.dll` to `/tmp/where-ever/it/extracts/extracted.dll`?

> distribute Node.js apps as single files.

Are you describing something like #27?  Or different?

**Reply by jviotti** [reply]
It must definitely be captured in https://github.com/nodejs/single-executable/pull/28. cc @RaisinTen

**Reply by zcbenz** [reply]
> Should this be included in a formal list of design requirements, for the VFS or for the SPA bootstrapper?

I think it should be part of VFS format.

The point of it is to not extract any DLL file during runtime, as this will be considered extremely suspicious by anti-virus softwares. So we can not have bootstrapper to extract the files on startup, the DLL files must be shipped as independent files together with the executable or the bundled package file, and yet be treated as part of the VFS.

> Are you describing something like https://github.com/nodejs/single-executable/discussions/27? Or different?

Yeah https://github.com/nodejs/single-executable/discussions/27 covers what I want.

**Reply by jviotti** [reply]
> The point of it is to not extract any DLL file during runtime, as this will be considered extremely suspicious by anti-virus softwares.

Yeah, this is spot on and a fact very much outside of our control. This is a hard constrain that means at least DLLs must be supported outside of the VFS.

**Reply by cspotcode** [reply]
> and yet be treated as part of the VFS.

What, precisely, should this mean?  Other discussions put the VFS at a location where it is *impossible* for files to exist on the physical filesystem, so the DLL's physical location cannot be a sub-path of the VFS.  Does the symlink idea from https://github.com/nodejs/single-executable/discussions/30#discussioncomment-3539403 fit the bill?

**Reply by jviotti** [reply]
@cspotcode I think Electron's ASAR get this right and we can take inspiration from it. The ASAR header is describes the files in the VFS and the corresponding offsets. Yet, the ASAR header is permitted to also point to files outside of the ASAR, siblings to where the VFS is stored.

**Reply by jviotti** [reply]
One interesting observation is that if we need to obey this hard constraint (at least for DLLs), the most of the archive formats discussed so far, including ZIP, are instantly out of the table, for better or worse

**Reply by cspotcode** [reply]
Does ASAR handle this by reserving a physical directory on disk, and mounting the ASAR onto that directory?  If yes, ZIP should work well.  If no, do we have a concrete example of all the paths in this scenario when using ASAR?  It will help with discussion.

- path to executable
- path to DLL
- path to ASAR
- VFS path of non-DLL within the ASAR

When ASAR points to a location outside of the archive, what does that mean for path strings within the SEA?  (for example `require.resolve('./native.node')`  What are those paths?  Are they non-existent on the physical FS?  Can native codepaths within nodejs understand them?  If the path doesn't exist on disk, how is a DLL loaded?  Or is it using virtual symlinks like suggested above?

One possible example of the bulleted list, maybe ASAR does something similar or different:

- path to executable: `/foo/executable`
- path to DLL: `/foo/data/lib.dll`
- path to ASAR: `/foo/data.asar`
- VFS path of non-DLL within the ASAR: `/foo/data/index.js`

In my hypothetical example, a physical directory on disk is used as the mount point for the VFS.  The contents of a ZIP and the physical contents on disk are overlaid.  This relies on reserving the `/foo/data` directory for exclusive use of this SEA, for example because it's in a good installation location.

This is a different solution to "No interference with valid paths in the file system."  Instead of "ensure paths never conflict," it is "have an installation directory."

**Reply by arcanis** [reply]
> One interesting observation is that if we need to obey this hard constraint (at least for DLLs), the most of the archive formats discussed so far, including ZIP, are instantly out of the table, for better or worse

Zip supports symlinks, which should make this possible, correct? DLLs can just be declared as symlinks to relative paths outside the archive.

**Reply by jviotti** [reply]
@cspotcode 

> Does ASAR handle this by reserving a physical directory on disk, and mounting the ASAR onto that directory?

No, it literally places the files outside of it at build time. The convention in Electron is something like this:

- `app.asar` (contains all your JS stuff)
- `app.asar.unpacked/path/to/library.dll` (a sibling full blown directory)

In this case, `app.asar` has a header declaring `path/to/library.dll`, stating that such file is "unpacked"

**Reply by jviotti** [reply]
@arcanis Interesting. I didn't know about it. Do you have any good resources I can use to read more about this?

**Reply by arcanis** [reply]
Zip has a field to store [external attributes](https://www.commandlinux.com/man-page/man3/zip_file_get_external_attributes.3.html). It's used to store the [`st_mode`](https://www.gnu.org/software/libc/manual/html_node/Testing-File-Type.html) stat field, which zip extractors check to see if it is set to `S_IFLNK`. If it is, the entry is treated as a symlink whose target is the path stored in the data segment.

**Reply by cspotcode** [reply]
> No, it literally places the files outside of it at build time.

I think we're talking about something similar, "This relies on reserving the /foo/data directory for exclusive use of this SEA, for example because it's in a good installation location."  But I was thinking that the asar VFS is mounted as an overlay on top of `app.asar.unpacked` or else the `app.asar.unpacked` is mounted as an overlay on top of `app.asar`.  Is either of those happening?

If an Electron app tries to `fs.readFile` or `fs.stat('/app.asar/path/to/library.dll')` and/or `fs.stat('/app.asar.unpacked/path/to/library.dll')`, what happens?  What about `/app.asar/file/within/thevfs.txt` or `/app.asar.unpacked/file/within/thevfs.txt`?

**Reply by zcbenz** [reply]
In Electron `/app.asar.unpacked` is not a special path, so both `fs.stat('/app.asar.unpacked/path/to/library.dll')` and  `/app.asar.unpacked/file/within/thevfs.txt` will try to access the real on the real filesystem.

The VFS in Electron is implemented by monkey-patching the `fs` module, so `fs.stat('/app.asar/path/to/library.dll')` will be translated to `fs.stat('/app.asar.unpacked/path/to/library.dll')` and it will return the stats of the real file, while `fs.stat('/app.asar/file/within/thevfs.txt')` will return a fake Stats object according to the information stored in `app.asar`.

**Reply by cspotcode** [reply]
> (virtual path) will be translated to (unpacked path)

How does that translation work?  Is it a virtual symlink, or something else?  For example, if I want to spawn an unpacked executable, how does my app obtain the unpacked path?  System calls will not understand the virtual path.  If it's a virtual symlink, then `fs.realpath` can give me the unpacked path.

**Reply by ljharb** [reply]
Why would we assume that SEAs are unpackable? This only needs to be a one-way process.

**Reply by RaisinTen** [reply]
> > (virtual path) will be translated to (unpacked path)
> 
> How does that translation work? Is it a virtual symlink, or something else?

It's a virtual symlink. The ASAR format has a field called `unpacked` for this purpose. If that's true for a file, the application can simply substitute the `app.asar` part in the path with `app.asar.unpacked` which is a directory on the real file system that contains the file on disk. You may take a look at https://github.com/electron/electron/blob/d027975cd7d1af8c892826aa3bdd0c6d7df7bb96/docs/tutorial/asar-archives.md#adding-unpacked-files-to-asar-archives. I'm pasting the contents here in case it gets lost because the [PR](https://github.com/electron/electron/pull/35563) that's adding the doc hasn't landed yet:

<details>

## Adding Unpacked Files to ASAR archives

As stated above, some Node APIs will unpack the file to the filesystem when
called. Apart from the performance issues, various anti-virus scanners might
be triggered by this behavior.

As a workaround, you can leave various files unpacked using the `--unpack` option.
In the following example, shared libraries of native Node.js modules will not be
packed:

```sh
$ asar pack app app.asar --unpack *.node
```

After running the command, you will notice that a folder named `app.asar.unpacked`
was created together with the `app.asar` file. It contains the unpacked files
and should be shipped together with the `app.asar` archive.

</details>

> For example, if I want to spawn an unpacked executable, how does my app obtain the unpacked path? System calls will not understand the virtual path. If it's a virtual symlink, then `fs.realpath` can give me the unpacked path.

The application would need to do the substitution I mentioned above before spawning the executable.

**Reply by RaisinTen** [reply]
> Why would we assume that SEAs are unpackable? This only needs to be a one-way process.

Technically it's not that these files are getting "unpacked", it's just that these aren't even included in the asar during creation, rather a parallel directory with a slightly different name gets created that contains the files that are not supposed to belong in the asar.

**Reply by cspotcode** [reply]
> the application can simply substitute the app.asar part in the path with app.asar.unpacked

I assume electron exposes an API to do this string manipulation?

**Reply by jviotti** [reply]
It is done implicit as part of their `fs` monkey-patching logic. @RaisinTen might have links to the relevant parts of the source code

**Reply by cspotcode** [reply]
*This was initially sent as an email reply, which made it appear in [the wrong place](https://github.com/nodejs/single-executable/discussions/30#discussioncomment-3609717).  I'm copying it here*

What about APIs that accept a string that is a path that are not the FS
API?  For example, worker_thread, spawn, fork, etc?  Since we're talking
about things like dlls, where the whole point is that those paths cannot be
VFS paths because of native code / OS requirements.

**Reply by zcbenz** [reply]
How Electron does monkey patching can be found at https://github.com/electron/electron/blob/main/lib/asar/fs-wrapper.ts#L236.

* For APIs like `fs.access` and `fs.exists`, fake information would be returned based on the manifest of ASAR file.
* For `fs.readFile`, it is reimplemented to read from ASAR file with `fs.read`.
* For `child_process.execXXX`, the file would be extracted first and then its path would be passed.
* If the passed path is an "unpacked" file, then the path to the real file would be passed instead of doing the above preprocessings.

**Reply by jviotti** [reply]
@zcbenz 

> For child_process.fork, the file would be extracted first and then its path would be passed.

Have you witnessed this being problematic (like with the DLLs) too?

**Reply by zcbenz** [reply]
Sorry I just re-read the code and found `fork` does not extract files since the child Node.js process created by Electron can recognize files in ASAR archives. Only `child_process.execXXX` extracts files.

**Comment by jviotti** [comment]
All very good point @zcbenz ! Thanks a lot for chiming in. We would love to have you in our calls if you can find the time (see https://github.com/nodejs/single-executable/discussions/19)

> On Windows, extracting DLLs and executing them is a red alarm to anti-virus softwares, so when supporting native addons with VFS there should be a way to leave the DLLs on the real filesystem, like the ["unpack" feature of ASAR](https://github.com/electron/asar#excluding-multiple-resources-from-being-packed).

Indeed! We have been extensibly discussing this on today's SEA call. Apart from AV concerns, this seems to be a good solution for making native add-ons that interact with I/O also play nicely with the VFS. See https://github.com/nodejs/single-executable/discussions/29 for some related discussions.

> On macOS executables are usually signed and the single executable implementation should be compatible with signing. Many single executable implementations are concatenating the packages to the tail of the executables, which would be broken after signing, but it can be fixed by https://github.com/vercel/pkg/pull/1164.

Yeah, this is a top concern. To solve this problem, we have created https://github.com/postmanlabs/postject, a tool to inject arbitrary data to pre-compiled executables as proper Mach-O/ELF/PE sections.

> While this repo is about single executable, it would be nice supporting reading from multiple bundled packages in one process, so it would be possible to distribute Node.js apps as single files. For example users can execute downloaded apps with something like node downloaded_app.nodeapp or by double clicking the package file, and app developers can do fork('/path/to/another/app.nodeapp').

This is a very interesting idea. We are talking about providing clear interfaces for extending `require()`, the `fs` module and others to avoid monkey-patching (like what Electron does with ASARs at the moment), which could in theory be used to implement something like this too? @RaisinTen @addaleax @arcanis

**Comment by arcanis** [comment]
> While this repo is about single executable, it would be nice supporting reading from multiple bundled packages in one process, so it would be possible to distribute Node.js apps as single files. For example users can execute downloaded apps with something like node downloaded_app.nodeapp or by double clicking the package file, and app developers can do fork('/path/to/another/app.nodeapp').

Yep, I think it's in line with my suggestion at https://github.com/nodejs/single-executable/discussions/27 (edit: Ah, I see someone mentioned it already 😆)

**Comment by cspotcode** [comment]
_EDIT This was meant to be a reply to a thread above, but Github's email integration doesn't do that.  I've copied it manually: https://github.com/nodejs/single-executable/discussions/30#discussioncomment-3610272_

What about APIs that accept a string that is a path that are not the FS
API?  For example, worker_thread, spawn, fork, etc?  Since we're talking
about things like dlls, where the whole point is that those paths cannot be
VFS paths because of native code / OS requirements.

**Reply by jviotti** [reply]
Yeah, these have to be taken care of. Projects like PKG and Electron monkey patch all of them. However, it is very tricky to maintain all these monkey-patches, which is the reason I was suggesting to experiment with providing a lower-level hooking capability to Node.js here: https://github.com/nodejs/single-executable/pull/31/files#diff-52c3ed092f4b8e8c123a7369c422f8846e288fe8abf3752b1d7eed4ec7808f83R33

**Reply by cspotcode** [reply]
I copied this to the place where I originally intended to post it: https://github.com/nodejs/single-executable/discussions/30#discussioncomment-3610272
